### PR TITLE
fix(megatron): make FP16 optimizer settings configurable

### LINKED ISSUE #145: [Task 15] Megatron FP16 optimizer 配置去硬编码方案确认
我认领了任务 15，把我对问题的理解、实现范围和验证口径记录在这里。Draft PR 会同步提交，后续有反馈我再继续调整。

## 我对问题的理解

我看了 `relax/backends/megatron/model.py` 里的 `setup_model_and_optimizer()`。现在只要开启 FP16，下面四个值就会在创建 `OptimizerConfig` 时被重新写死：

```text
initial_loss_scale = 32768
min_loss_scale = 1
use_precision_aware_optimizer = True
store_param_remainders = False
```

所以即使用户从命令行传了其他值，最后也不会生效。前三项在当前 Megatron parser 中已经有对应参数，`store_param_remainders` 还没有从 Relax 的命令行暴露出来。

## 我准备这样改

- 我会复用 Megatron 现有的 `initial_loss_scale`、`min_loss_scale` 和 `use_precision_aware_optimizer` 参数，并补上 `--no-use-precision-aware-optimizer`。
- 我会增加 `--store-param-remainders` 和 `--no-store-param-remainders`，这样开启和关闭都能明确指定。
- 这四项在用户没有设置时先保留为 `None`，然后在 Megatron 校验前补上对应模式的默认值。
- FP16 缺参时继续沿用 Relax 现在的行为：`32768 / 1 / 开启 precision-aware / 关闭 param-remainders`，同时只输出一条 warning，把实际采用的参数列出来。
- BF16 等非 FP16 路径继续使用当前 Megatron 的默认值，不输出 FP16 fallback warning。
- 我会删除 `model.py` 中四项无条件覆盖，让显式传入的值原样进入 `OptimizerConfig`。
- 我会在 `scripts/training/text/run-qwen3-4B-fp16-8xgpu.sh` 中写明四项配置，并把脚本收到的额外参数放在命令末尾，方便用户覆盖脚本默认值。
- 中英文配置文档会一起更新。

对于 loss scale，我准备只做必要校验：两个值都必须是大于 0 的有限数值，并且 `min_loss_scale <= initial_loss_scale`。我不会额外限制为 2 的幂，因为当前 Megatron 接受任意正浮点数。

## 我准备怎么验证

```bash
python3 -m pytest -v tests/backends/megatron/test_fp16_optimizer_config.py
python3 -m pytest -q tests/backends/megatron
bash -n scripts/training/text/run-qwen3-4B-fp16-8xgpu.sh
pre-commit run --all-files
```

我会覆盖未配置、部分配置、全部显式配置、脚本参数覆盖、非法值、重复归一化不重复 warning、BF16 默认值，以及真实 `OptimizerConfig` 构造。

我的验证基线是 `cfda05351d4b137f60e2b71d048beab1316d9d9f`，使用官方镜像 `ghcr.io/redai-infra/relaxrl:dev-20260715-8325919e`，环境为 Python 3.12.3、PyTorch 2.11.0+cu129、CUDA 12.9.1 和 8 张 A100 80GB。

当前机器上没有这个脚本需要的 Qwen3-4B 模型、Megatron checkpoint 和 DAPO 数据，所以本次计划使用真实 Megatron parser、真实 `OptimizerConfig`、Megatron backend 回归和脚本检查作为验证证据。

## 想确认三点

1. 我把上面四项都放在任务 15 的修改范围内，这样理解是否正确？
2. 目前每个 arguments 实例最多输出一次缺参 warning，正常运行时通常对应每个进程一次。这个语义是否可以，还是要求只在分布式 rank 0 输出？
3. 在缺少完整训练资产的情况下，上面的 parser、`OptimizerConfig` 和 backend 回归是否足够作为验收证据？

### PR DESCRIPTION
**Pull request by huxy1225-cloud**
## Summary

In FP16 mode, Megatron model setup currently overwrites four optimizer values after CLI parsing, so explicit user values cannot take effect. I removed those overrides and made the four settings configurable from the command line.

When an FP16 option is omitted, I keep the existing Relax value as a compatibility fallback and report the applied fallback in one warning. I also updated the Qwen3-4B FP16 recipe, added focused tests, and documented the behavior in both English and Chinese.

Closes #145

Related to #86 (Task 15).

## Changes

- I preserved an unset state for `initial_loss_scale`, `min_loss_scale`, `use_precision_aware_optimizer`, and `store_param_remainders`, so an omitted value can be distinguished from an explicit value.
- I added the missing negative/positive boolean flags.
- I resolve FP16 compatibility fallbacks before Megatron validation instead of overwriting values in `model.py`.
- I kept the pinned Megatron defaults for non-FP16 modes and added validation for invalid dynamic loss-scale values.
- I made the target recipe pass all four values explicitly and append `"$@"`, so later user arguments can override the recipe.
- I added focused FP16/BF16 tests and updated both configuration guides.

## Verification

I ran the following checks in the official Relax image (`ghcr.io/redai-infra/relaxrl:dev-20260715-8325919e`, Python 3.12.3, PyTorch 2.11.0+cu129, CUDA 12.9.1):

```bash
python3 -m pytest -q tests/backends/megatron/test_fp16_optimizer_config.py
# 26 passed

python3 -m pytest -q tests/backends/megatron
# 112 passed, 1 skipped

bash -n scripts/training/text/run-qwen3-4B-fp16-8xgpu.sh
# passed

pre-commit run --all-files
# all hooks passed

git diff --check origin/main..HEAD
# passed
```

The focused tests use the pinned Megatron parser and construct real FP16 and BF16 `OptimizerConfig` objects.

The repository's CPU-only GitHub jobs do not install Megatron, so this focused module follows the existing optional-dependency pattern and skips there. All 26 focused tests run in the official image above.

Full Qwen3-4B training was not run because this environment does not contain the required model, Megatron checkpoint, or DAPO dataset. Verification for this PR is limited to the configuration path and related backend regression tests.

## Risk & Rollback

The main risk is changing behavior when users omit these options. I kept the old Relax FP16 values for that case and added separate tests for fallback, explicit values, and BF16 defaults.

This change can be rolled back by reverting the commit; no data migration is involved.

## Checklist

- [x] I kept the diff limited to task 15.
- [x] I added tests for omitted, partial, explicit, invalid, FP16, and BF16 cases.
- [x] I updated the English and Chinese documentation together.
- [x] I ran the related Megatron backend tests.
- [x] I ran the full pre-commit suite.
- [x] I linked clarification Issue #145.
- [x] GitHub CI has passed.
- [ ] I have addressed all blocking review comments.

### CODE REVIEW

File: relax/backends/megatron/arguments.py [hunk 1]
```diff
@@ -21,6 +22,84 @@
 
 logger = get_logger(__name__)
 
+_FP16_OPTIMIZER_FALLBACKS = {
```
Comment (NINGBENZHE) [inline relax/backends/megatron/arguments.py:26]: 这里为四个字段维护了多张平行配置表和独立 formatter，后续字段调整容易不同步。建议只保留 Relax 的 FP16 compatibility fallback；非 FP16 默认值直接从 OptimizerConfig 获取，option 名称统一生成或合并到一张规格表，减少重复和默认值漂移。

File: relax/backends/megatron/model.py [hunk 2]
```diff
@@ -224,6 +225,21 @@ def get_optimizer_param_scheduler(args: Namespace, optimizer: MegatronOptimizer)
     return opt_param_scheduler
 
 
+def _build_optimizer_config_kwargs(args: Namespace) -> dict[str, object]:
+    """Build optimizer kwargs from normalized runtime arguments."""
+    _resolve_optimizer_precision_args(args)
```
Comment (NINGBENZHE) [inline relax/backends/megatron/model.py]: _resolve_optimizer_precision_args() 已在参数解析阶段调用，这里再次执行会让模型构造同时承担 fallback、校验和 warning 职责。建议参数层只归一化一次，_build_optimizer_config_kwargs() 只负责从已归一化的 args 构造 kwargs，减少耦合和重复调用。

File: relax/backends/megatron/arguments.py [hunk 3]
```diff
@@ -21,6 +22,84 @@
...
+
+def _format_optimizer_fallback(name: str, value: object) -> str:
+    boolean_options = _FP16_OPTIMIZER_BOOLEAN_OPTIONS.get(name)
+    if boolean_options is not None:
+        return boolean_options[value]
+    return f"{_FP16_OPTIMIZER_OPTIONS[name]} {value!r}"
+
+
+def _validate_fp16_optimizer_args(args) -> None:
+    for name in ("initial_loss_scale", "min_loss_scale"):
```
Comment (NINGBENZHE) [inline relax/backends/megatron/arguments.py]: 这里没有区分静态 args.loss_scale。设置 --loss-scale 后，Megatron 不会使用 initial_loss_scale 和 min_loss_scale，因此不应再校验或提示这两个动态参数缺失。建议仅在 args.loss_scale is None 时处理动态 loss-scale 配置。

**Review by NINGBENZHE — CHANGES_REQUESTED** [review]

File: relax/backends/megatron/arguments.py (hunk 1 above)
Comment (HuangShengZeBlueSky) [reply relax/backends/megatron/arguments.py:26]: ​

File: relax/backends/megatron/model.py (hunk 2 above)
Comment (HuangShengZeBlueSky) [reply relax/backends/megatron/model.py]: ​

File: relax/backends/megatron/arguments.py (hunk 3 above)
Comment (HuangShengZeBlueSky) [reply relax/backends/megatron/arguments.py]: ​

**Comment by HuangShengZeBlueSky** [comment]
​

File: relax/backends/megatron/arguments.py (hunk 1 above)
Comment (huxy1225-cloud) [reply relax/backends/megatron/arguments.py:26]: 已按建议收敛配置：现在只保留 Relax 的 `_FP16_OPTIMIZER_FALLBACKS`，非 FP16 默认值直接从 `OptimizerConfig` 读取，CLI option 名称统一生成，不再维护多张平行默认值表。

File: relax/backends/megatron/model.py (hunk 2 above)
Comment (huxy1225-cloud) [reply relax/backends/megatron/model.py]: 已删除 model 层对 `_resolve_optimizer_precision_args()` 的重复调用。现在只在参数层完成归一化，`_build_optimizer_config_kwargs()` 只读取 args 构造 kwargs；同时补了不修改 args、也不重复发 warning 的测试。

File: relax/backends/megatron/arguments.py (hunk 3 above)
Comment (huxy1225-cloud) [reply relax/backends/megatron/arguments.py]: 已区分静态和动态 loss scale。设置 `args.loss_scale` 时，不再补齐、校验或提示 `initial_loss_scale` / `min_loss_scale`；只有动态模式才处理这两个字段，并补了静态 loss scale 的覆盖测试。

**Comment by huxy1225-cloud** [comment]
review 修改已更新到 `3c2c4a6`，三条意见均已逐项处理。

验证结果：
- FP16 optimizer 聚焦测试：29 passed
- Megatron 后端回归：115 passed, 1 skipped
- ruff check / format：通过
- 训练脚本语法与 git diff 检查：通过

麻烦再帮忙看一下。

**Review by NINGBENZHE — APPROVED** [review]
lgtm
